File: frontend/crm3/src/stores/store.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

export interface Store {
  id: number
  name: string
  address?: string
  phone?: string
  email?: string
  status: 'active' | 'inactive' | 'maintenance'
  manager_id?: number
  created_at: string
  updated_at: string
}

export interface StoreCreateRequest {
  name: string
  address?: string
  phone?: string
  email?: string
  status: 'active' | 'inactive' | 'maintenance'
  manager_id?: number
}

export interface StoreUpdateRequest {
  name: string
  address?: string
  phone?: string
  email?: string
  status: 'active' | 'inactive' | 'maintenance'
  manager_id?: number
}

export interface StoreSearchParams {
  name?: string
  status?: string
  page?: number
  limit?: number
}

export interface PaginatedStoreResponse {
  stores: Store[]
  total: number
  page: number
  limit: number
  totalPages: number
}

export const useStoreStore = defineStore('store', () => {
  // State
  const stores = ref<Store[]>([])
  const currentStore = ref<Store | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)
  const total = ref(0)
  const currentPage = ref(1)
  const itemsPerPage = ref(10)

  // Computed
  const totalPages = computed(() => Math.ceil(total.value / itemsPerPage.value))

  const activeStores = computed(() =>
    stores.value.filter(store => store.status === 'active')
  )

  const inactiveStores = computed(() =>
    stores.value.filter(store => store.status === 'inactive')
  )

  const maintenanceStores = computed(() =>
    stores.value.filter(store => store.status === 'maintenance')
  )

  const statusCounts = computed(() => ({
    active: activeStores.value.length,
    inactive: inactiveStores.value.length,
    maintenance: maintenanceStores.value.length,
    total: stores.value.length
  }))

  // Actions
  const fetchStores = async (params: StoreSearchParams = {}) => {
    loading.value = true
    error.value = null

    try {
      // デフォルトパラメータ
      const queryParams = {
        page: params.page || currentPage.value,
        limit: params.limit || itemsPerPage.value,
        ...params
      }

      const query = new URLSearchParams()
      Object.entries(queryParams).forEach(([key, value]) => {
        if (value !== undefined && value !== null && value !== '') {
          query.append(key, value.toString())
        }
      })

      const response = await fetch(`/api/stores?${query}`, {
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('token')}`,
          'Content-Type': 'application/json',
        },
      })

      if (!response.ok) {
        const errorData = await response.json()
        throw new Error(errorData.error || 'Failed to fetch stores')
      }

      const data: PaginatedStoreResponse = await response.json()

      stores.value = data.stores || []
      total.value = data.total
      currentPage.value = data.page

    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to fetch stores'
      stores.value = []
      total.value = 0
    } finally {
      loading.value = false
    }
  }

  const fetchStore = async (id: number) => {
    loading.value = true
    error.value = null

    try {
      const response = await fetch(`/api/stores/${id}`, {
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('token')}`,
          'Content-Type': 'application/json',
        },
      })

      if (!response.ok) {
        const errorData = await response.json()
        throw new Error(errorData.error || 'Failed to fetch store')
      }

      const data: Store = await response.json()
      currentStore.value = data
      return data

    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to fetch store'
      currentStore.value = null
      throw err
    } finally {
      loading.value = false
    }
  }

  const createStore = async (storeData: StoreCreateRequest) => {
    loading.value = true
    error.value = null

    try {
      const response = await fetch('/api/stores', {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('token')}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(storeData),
      })

      if (!response.ok) {
        const errorData = await response.json()
        throw new Error(errorData.error || 'Failed to create store')
      }

      const newStore: Store = await response.json()

      // 現在の一覧に追加
      stores.value.unshift(newStore)
      total.value += 1

      return newStore

    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to create store'
      throw err
    } finally {
      loading.value = false
    }
  }

  const updateStore = async (id: number, storeData: StoreUpdateRequest) => {
    loading.value = true
    error.value = null

    try {
      const response = await fetch(`/api/stores/${id}`, {
        method: 'PUT',
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('token')}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(storeData),
      })

      if (!response.ok) {
        const errorData = await response.json()
        throw new Error(errorData.error || 'Failed to update store')
      }

      const updatedStore: Store = await response.json()

      // 一覧内の店舗を更新
      const index = stores.value.findIndex(store => store.id === id)
      if (index !== -1) {
        stores.value[index] = updatedStore
      }

      // 現在の店舗も更新
      if (currentStore.value && currentStore.value.id === id) {
        currentStore.value = updatedStore
      }

      return updatedStore

    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to update store'
      throw err
    } finally {
      loading.value = false
    }
  }

  const deleteStore = async (id: number) => {
    loading.value = true
    error.value = null

    try {
      const response = await fetch(`/api/stores/${id}`, {
        method: 'DELETE',
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('token')}`,
          'Content-Type': 'application/json',
        },
      })

      if (!response.ok) {
        const errorData = await response.json()
        throw new Error(errorData.error || 'Failed to delete store')
      }

      // 一覧から削除
      const index = stores.value.findIndex(store => store.id === id)
      if (index !== -1) {
        stores.value.splice(index, 1)
        total.value -= 1
      }

      // 現在の店舗をクリア
      if (currentStore.value && currentStore.value.id === id) {
        currentStore.value = null
      }

    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to delete store'
      throw err
    } finally {
      loading.value = false
    }
  }

  const updateStoreStatus = async (id: number, status: Store['status']) => {
    loading.value = true
    error.value = null

    try {
      const response = await fetch(`/api/stores/${id}/status`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('token')}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ status }),
      })

      if (!response.ok) {
        const errorData = await response.json()
        throw new Error(errorData.error || 'Failed to update store status')
      }

      const updatedStore: Store = await response.json()

      // 一覧内の店舗を更新
      const index = stores.value.findIndex(store => store.id === id)
      if (index !== -1) {
        stores.value[index] = updatedStore
      }

      // 現在の店舗も更新
      if (currentStore.value && currentStore.value.id === id) {
        currentStore.value = updatedStore
      }

      return updatedStore

    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to update store status'
      throw err
    } finally {
      loading.value = false
    }
  }

  const searchStores = async (searchParams: StoreSearchParams) => {
    currentPage.value = 1 // 検索時はページをリセット
    await fetchStores(searchParams)
  }

  const setPage = async (page: number) => {
    currentPage.value = page
    await fetchStores()
  }

  const setItemsPerPage = async (limit: number) => {
    itemsPerPage.value = limit
    currentPage.value = 1 // ページサイズ変更時はページをリセット
    await fetchStores()
  }

  const clearError = () => {
    error.value = null
  }

  const clearCurrentStore = () => {
    currentStore.value = null
  }

  const getStoreStatusLabel = (status: Store['status']): string => {
    const statusLabels: Record<Store['status'], string> = {
      active: '営業中',
      inactive: '休業中',
      maintenance: 'メンテナンス中'
    }
    return statusLabels[status] || status
  }

  const getStoreStatusColor = (status: Store['status']): string => {
    const statusColors: Record<Store['status'], string> = {
      active: 'success',
      inactive: 'warning',
      maintenance: 'error'
    }
    return statusColors[status] || 'default'
  }

  return {
    // State
    stores,
    currentStore,
    loading,
    error,
    total,
    currentPage,
    itemsPerPage,

    // Computed
    totalPages,
    activeStores,
    inactiveStores,
    maintenanceStores,
    statusCounts,

    // Actions
    fetchStores,
    fetchStore,
    createStore,
    updateStore,
    deleteStore,
    updateStoreStatus,
    searchStores,
    setPage,
    setItemsPerPage,
    clearError,
    clearCurrentStore,
    getStoreStatusLabel,
    getStoreStatusColor
  }
})

File: backend/crm3_api/models/store.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Store 店舗モデル
type Store struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	Name      string    `json:"name" gorm:"not null" validate:"required,min=1,max=100"`
	Address   *string   `json:"address,omitempty" validate:"omitempty,max=500"`
	Phone     *string   `json:"phone,omitempty" validate:"omitempty,max=20"`
	Email     *string   `json:"email,omitempty" validate:"omitempty,email,max=320"`
	Status    string    `json:"status" gorm:"not null" validate:"required,oneof=active inactive maintenance"`
	ManagerID *uint     `json:"manager_id,omitempty" gorm:"index"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	// 関連
	Manager *User `json:"manager,omitempty" gorm:"foreignKey:ManagerID"`
}

// StoreCreateRequest 店舗作成リクエスト
type StoreCreateRequest struct {
	Name      string  `json:"name" validate:"required,min=1,max=100"`
	Address   *string `json:"address,omitempty" validate:"omitempty,max=500"`
	Phone     *string `json:"phone,omitempty" validate:"omitempty,max=20"`
	Email     *string `json:"email,omitempty" validate:"omitempty,email,max=320"`
	Status    string  `json:"status" validate:"required,oneof=active inactive maintenance"`
	ManagerID *uint   `json:"manager_id,omitempty"`
}

// StoreUpdateRequest 店舗更新リクエスト
type StoreUpdateRequest struct {
	Name      string  `json:"name" validate:"required,min=1,max=100"`
	Address   *string `json:"address,omitempty" validate:"omitempty,max=500"`
	Phone     *string `json:"phone,omitempty" validate:"omitempty,max=20"`
	Email     *string `json:"email,omitempty" validate:"omitempty,email,max=320"`
	Status    string  `json:"status" validate:"required,oneof=active inactive maintenance"`
	ManagerID *uint   `json:"manager_id,omitempty"`
}

// StoreStatusUpdateRequest 店舗ステータス更新リクエスト
type StoreStatusUpdateRequest struct {
	Status string `json:"status" validate:"required,oneof=active inactive maintenance"`
}

// StoreSearchParams 店舗検索パラメータ
type StoreSearchParams struct {
	Name     string `query:"name"`
	Status   string `query:"status"`
	Page     int    `query:"page"`
	Limit    int    `query:"limit"`
	ManagerID *uint  `query:"manager_id"`
}

// PaginatedStoreResponse ページネーション付き店舗レスポンス
type PaginatedStoreResponse struct {
	Stores     []Store `json:"stores"`
	Total      int64   `json:"total"`
	Page       int     `json:"page"`
	Limit      int     `json:"limit"`
	TotalPages int     `json:"total_pages"`
}

// StoreStatusCounts 店舗ステータス別件数
type StoreStatusCounts struct {
	Active      int64 `json:"active"`
	Inactive    int64 `json:"inactive"`
	Maintenance int64 `json:"maintenance"`
	Total       int64 `json:"total"`
}

// ToStore StoreCreateRequestをStoreに変換
func (req *StoreCreateRequest) ToStore() Store {
	return Store{
		Name:      req.Name,
		Address:   req.Address,
		Phone:     req.Phone,
		Email:     req.Email,
		Status:    req.Status,
		ManagerID: req.ManagerID,
	}
}

// ToStore StoreUpdateRequestをStoreに変換
func (req *StoreUpdateRequest) ToStore() Store {
	return Store{
		Name:      req.Name,
		Address:   req.Address,
		Phone:     req.Phone,
		Email:     req.Email,
		Status:    req.Status,
		ManagerID: req.ManagerID,
	}
}

// BeforeCreate 作成前のフック
func (s *Store) BeforeCreate(tx *gorm.DB) error {
	// 店舗名の重複チェック
	var count int64
	if err := tx.Model(&Store{}).Where("name = ?", s.Name).Count(&count).Error; err != nil {
		return err
	}
	if count > 0 {
		return gorm.ErrDuplicatedKey
	}

	// 管理者の存在チェック（指定されている場合）
	if s.ManagerID != nil {
		var manager User
		if err := tx.First(&manager, *s.ManagerID).Error; err != nil {
			if err == gorm.ErrRecordNotFound {
				return gorm.ErrInvalidValue
			}
			return err
		}
	}

	return nil
}

// BeforeUpdate 更新前のフック
func (s *Store) BeforeUpdate(tx *gorm.DB) error {
	// 店舗名の重複チェック（自分以外）
	var count int64
	if err := tx.Model(&Store{}).Where("name = ? AND id != ?", s.Name, s.ID).Count(&count).Error; err != nil {
		return err
	}
	if count > 0 {
		return gorm.ErrDuplicatedKey
	}

	// 管理者の存在チェック（指定されている場合）
	if s.ManagerID != nil {
		var manager User
		if err := tx.First(&manager, *s.ManagerID).Error; err != nil {
			if err == gorm.ErrRecordNotFound {
				return gorm.ErrInvalidValue
			}
			return err
		}
	}

	return nil
}

// GetStatusLabel ステータスラベルを取得
func (s *Store) GetStatusLabel() string {
	statusLabels := map[string]string{
		"active":      "営業中",
		"inactive":    "休業中",
		"maintenance": "メンテナンス中",
	}
	if label, exists := statusLabels[s.Status]; exists {
		return label
	}
	return s.Status
}

// IsActive 営業中かどうかを判定
func (s *Store) IsActive() bool {
	return s.Status == "active"
}

// IsInactive 休業中かどうかを判定
func (s *Store) IsInactive() bool {
	return s.Status == "inactive"
}

// IsUnderMaintenance メンテナンス中かどうかを判定
func (s *Store) IsUnderMaintenance() bool {
	return s.Status == "maintenance"
}

// CanBeActivated 営業開始可能かどうかを判定
func (s *Store) CanBeActivated() bool {
	return s.Status == "inactive" || s.Status == "maintenance"
}

// CanBeDeactivated 休業可能かどうかを判定
func (s *Store) CanBeDeactivated() bool {
	return s.Status == "active"
}

// CanGoToMaintenance メンテナンス状態に移行可能かどうかを判定
func (s *Store) CanGoToMaintenance() bool {
	return s.Status == "active" || s.Status == "inactive"
}

// ValidateStatusTransition ステータス遷移の妥当性をチェック
func (s *Store) ValidateStatusTransition(newStatus string) error {
	switch newStatus {
	case "active":
		if !s.CanBeActivated() {
			return gorm.ErrInvalidValue
		}
	case "inactive":
		if !s.CanBeDeactivated() {
			return gorm.ErrInvalidValue
		}
	case "maintenance":
		if !s.CanGoToMaintenance() {
			return gorm.ErrInvalidValue
		}
	default:
		return gorm.ErrInvalidValue
	}
	return nil
}

// UpdateStatus ステータスを更新
func (s *Store) UpdateStatus(newStatus string) error {
	if err := s.ValidateStatusTransition(newStatus); err != nil {
		return err
	}
	s.Status = newStatus
	return nil
}

// GetStatusColor ステータス色を取得
func (s *Store) GetStatusColor() string {
	statusColors := map[string]string{
		"active":      "success",
		"inactive":    "warning",
		"maintenance": "error",
	}
	if color, exists := statusColors[s.Status]; exists {
		return color
	}
	return "default"
}